amaru/src/consensus/header_validation.rs

```diff
@@ -0,0 +1,179 @@
+    }
+}
+
+pub struct Worker {}
+
+#[async_trait::async_trait(?Send)]
+impl gasket::framework::Worker<Stage> for Worker {
+    async fn bootstrap(_stage: &Stage) -> Result<Self, WorkerError> {
+        Ok(Self {})
+    }
+
+    async fn schedule(
+        &mut self,
+        stage: &mut Stage,
+    ) -> Result<WorkSchedule<PullEvent>, WorkerError> {
+        let unit = stage.upstream.recv().await.or_panic()?;
+
+        Ok(WorkSchedule::Unit(unit.payload))
+    }
+
+    async fn execute(&mut self, unit: &PullEvent, stage: &mut Stage) -> Result<(), WorkerError> {
+        match unit {
+            PullEvent::RollForward(point, raw_header) => {
+                info!(?point, "validating roll forward");
+                let header = MultiEraHeader::decode(6, None, raw_header)
+                    .map_err(|e| miette!(e))
+                    .or_panic()?;
+
+                match header {
+                    MultiEraHeader::BabbageCompatible(_) => {
+                        let minted_header = header.as_babbage().unwrap();
+                        let epoch = epoch_for_preprod_slot(minted_header.header_body.slot);
+
+                        // TODO: This is awkward, and should probably belong to the LedgerState
+                        // abstraction? The ledger shall keep track of the rolling nonce and
+                        // provide some endpoint for the consensus to access it.
+                        let epoch_nonce = stage
+                            .epoch_to_nonce
+                            .get(&epoch)
+                            .ok_or(miette!("epoch nonce not found"))
+                            .or_panic()?;
+
+                        let active_slots_coeff: FixedDecimal =
+                            FixedDecimal::from(5u64) / FixedDecimal::from(100u64);
+                        let c = (FixedDecimal::from(1u64) - active_slots_coeff).ln();
+                        let ledger = stage.ledger.lock().unwrap();
+                        let block_validator =
+                            BlockValidator::new(minted_header, &*ledger, epoch_nonce, &c);
+                        block_validator.validate().or_panic()?;
+                        info!(?minted_header.header_body.block_number, "validated block");
+                    }
+                    MultiEraHeader::ShelleyCompatible(_) => {
+                        trace!("shelley compatible header, skipping validation");
+                    }
+                    MultiEraHeader::EpochBoundary(_) => {
+                        trace!("epoch boundary header, skipping validation");
+                    }
+                    MultiEraHeader::Byron(_) => {
+                        trace!("byron header, skipping validation");
+                    }
+                }
+
+                let block = {
+                    info!("fetching block...");
+                    let mut peer_session = stage.peer_session.lock().unwrap();
+                    let client = (*peer_session).blockfetch();
```

Comment:
In "reality", we don't always fetch blocks, we only do so if we can extend our best chain with the retrieved header IIRC. Should there not be a separate "actor" to fetch blocks?

Reply:
This is what @AndrewWestberg is hinting at in the module comments I believe. At this point, the validation worker is pulling the block directly; but we do want to move that into a separate stage of the pipeline.

Reply:
I left an empty Stage for moving this to a fetch module. However, I didn't think I could get all that done before I needed to leave next week for Dubai and get it properly handed off to @KtorZ .